### ISSUE TITLE
Two forms with Provider problem

### ISSUE THREAD

**Issue by tonoslav** [react-hook-form/react-hook-form#2076]
**Is your feature request related to a problem? Please describe.**
I am using **FormProvider** with **useFormContext** in my whole app, and now I am working on new form, where there is \<select\> where you choose from partners or add new in modal, now I have one main form and another form in modal.
Everything works good, but when I submit modal form, main form submits too. 
This is because of FormProvider I think because it (your code) use just one global context.
I have to use context because of my custom form components.
But both have own **useForm** like this
![image](https://user-images.githubusercontent.com/3399021/86541266-c3475980-bf0b-11ea-9859-5e225b79c8ac.png)

**Describe the solution you'd like**
Not sure why this happening even how to solve it, I think this should be automatic

**Describe alternatives you've considered**
Use another package? don't want to 

**Additional context**
You can see it here, I submit "top" form in modal and it shows error in bg (also submits it)
![partner](https://user-images.githubusercontent.com/3399021/86541205-6055c280-bf0b-11ea-9da7-7eef8712bf6e.gif)

Same in version 5 and 6 too

**Comment by bluebill1049** [comment]
i need more context on this one and a codesandbox.

**Comment by tonoslav** [comment]
@bluebill1049  I can try to make one tomorrow, but just try to create one form in another, submit that inside form , top form will submit too
Just to state , real html code is like

```html
<form id="main">
...
</form>

<div class="modal">
    <form id="modal-form-in-main-component">
     ..
     </form>
</div> 
```

and when I submit form#modal-form-in-main-component , also form#main is submited

**Comment by bluebill1049** [comment]
oh, I see the problem, you will need separate hook :) each hook instance contains registered input and attaches with handleSbumit.

**Comment by tonoslav** [comment]
@bluebill1049  but each of them use **useForm** in it  

So I have main component with main form and with call 

```tsx
// Prepare data
    const methods = useForm<formData>({
        defaultValues: props.data
    })

    // Prepare handler
    const onSubmit = useCallback(methods.handleSubmit((data: formData) => {
        console.log(data)
    }), [])

    // Render
    return (
        <Page title="New Invoice">
            <Form onSubmit={onSubmit} hideButton>
                <FormProvider {...methods}>
                 ...
```

and there is another other component in those "three dots" in which I have another component with form with **useForm**

```tsx
 // Prepare data
    const methods = useForm<formData>()
 
    // Prepare handler
    const onSubmit = useCallback(methods.handleSubmit(newData => {
        ...
    }), [])

    // Render
    return (
        <Modal title="New partner" onHide={onHide}>
            <Form onSubmit={onSubmit} hideButton>
                <FormProvider {...methods}>
                    <ModalBody>
                    ....
```

and when I submit this second form , it will also submit the first, this second form is in the first component but result html is in other element out of the first, so they don't submit other because of basic html feautures but because of something in react-hook-form, anything I can change what could help ?

**Form** component in those examples is just basic \<form tag

**Comment by bluebill1049** [comment]
you will have to reproduce in a simple codesandbox, i will take a look again :)

**Comment by xiaoyu-tamu** [comment]
I solved it with something like 
```
  const formId = React.useMemo(() => Math.random().toString(36).substr(2, 9), []);
<Form id={formId} />
<Button form={formId} type="submit"  />
```

**Comment by bluebill1049** [comment]
is this issue still exist?

**Comment by tonoslav** [comment]
@bluebill1049 I will close it , give me a day :)

**Comment by bluebill1049** [comment]
@tonoslav i am not rushing to close it, just a quick follow up. let me know if anything i can help too.

**Comment by tonoslav** [comment]
@bluebill1049 @xiaoyu-tamu 
Didn't help :( As Ive said I think it is not HTML problem , rather plugin/package

1st arrow with square points where main html form is
Other arrows shows evidence that Ive used **id** and **form** props 

Also main form has it with different name 

![Screenshot_1](https://user-images.githubusercontent.com/3399021/86846279-f5db8880-c0aa-11ea-9f3c-2e5d09f45fa4.png)

You can see errors are visible on both forms even I clicked only "add" button in modal

**Comment by bluebill1049** [comment]
> @bluebill1049 @xiaoyu-tamu
> Didn't help :( As Ive said I think it is not HTML problem , rather plugin/package
> 
> 1st arrow with square points where main html form is
> Other arrows shows evidence that Ive used **id** and **form** props
> 
> Also main form has it with different name
> 
> ![Screenshot_1](https://user-images.githubusercontent.com/3399021/86846279-f5db8880-c0aa-11ea-9f3c-2e5d09f45fa4.png)
> 
> You can see errors are visible on both forms even I clicked only "add" button in modal

Yes, it's not an HTML problem at all. Like I suggested above, you are trying to share useForm through context, but you should have separate useForm for each form.

**Comment by xiaoyu-tamu** [comment]
Try wrap you own onSubmit handler for the inner form, stop the propagation and pass event to form.handlerSubmit(()=>{})(event)

**Comment by tonoslav** [comment]
> Try this for your modal form
> 
> ```
> <form
>   id={formId}
>   name={formId}
>   noValidate
>   onSubmit={(event) => {
>     // see here
>     event?.stopPropagation();
>     form.handleSubmit(async (values, event) => {})(event);
>   }}
> >
>    {children}
> </form>
> ```
> 
> ![image](https://user-images.githubusercontent.com/33362998/86857757-ca45a780-c084-11ea-810c-d85e4f55b739.png)

This works, but I think this is kinda hack rather than maybe package problem

**Comment by bluebill1049** [comment]
Honestly, I don't think this is lib related issue. there is no magic on how we submit the form, in fact onSubmit is a react event.

**Comment by tonoslav** [comment]
@bluebill1049 but why would it submit form which is not within that other form? 
Imagine you have two forms next to each other

```html
<form id="one">...</form>
<form id="two">...</form>
```

and you click submit button in **form#two** , 
do you think **form#one** submits too or not?

**Comment by bluebill1049** [comment]
> @bluebill1049 but why would it submit form which is not within that other form?
> Imagine you have two forms next to each other
> 
> ```
> <form id="one">...</form>
> <form id="two">...</form>
> ```
> 
> and you click submit button in **form#two** ,
> do you think **form#one** submits too or not?

Why you are using the same hook instance for two forms? `useForm` was support to be per form, that wouldn't case issue like this. If you scroll up, I am keeping saying that but kinda get ignored. I think that's the root cause of the problem here.

**Comment by tonoslav** [comment]
@bluebill1049  I don't use the same hook I reacted to it 

![image](https://user-images.githubusercontent.com/3399021/86905671-b34d9680-c112-11ea-9ff8-c875572cdb73.png)

![image](https://user-images.githubusercontent.com/3399021/86905703-c1031c00-c112-11ea-8c2e-e9506a10c101.png)

**Comment by bluebill1049** [comment]
> @bluebill1049 I don't use the same hook I reacted to it
> 
> ![image](https://user-images.githubusercontent.com/3399021/86905671-b34d9680-c112-11ea-9ff8-c875572cdb73.png)
> 
> ![image](https://user-images.githubusercontent.com/3399021/86905703-c1031c00-c112-11ea-8c2e-e9506a10c101.png)

are you sharing `FormProvider`? or `FormProvider` each form?

**Comment by bluebill1049** [comment]
can you take a look at this CSB: https://codesandbox.io/s/dawn-field-t12y1?file=/src/App.js?

**Comment by tonoslav** [comment]
@bluebill1049  @xiaoyu-tamu ok I think **xiaoyu's solution is right** because I am using **createPortal** to render element in modal, but for events as stated on react docs, element kinda stays where it was created, so even it doesn't looks like I have two forms in each other but I have :D 

Code interpretation
```tsx
<Form id="main">
    ....
    <Form id="create-portal-to-modal">
     ....
    </Form>
</Form>
```
results to html
```html
<form id="main">...</form>
<div class="modal"><form id="create-portal-to-modal">...</form></div>
```

But even it is in HTML out of main form, in react tree it is still in the main and submit event bubble right into main form.

![image](https://user-images.githubusercontent.com/3399021/86911126-10e5e100-c11b-11ea-889d-155f2e6d9765.png)
https://reactjs.org/docs/portals.html#event-bubbling-through-portals

**TLDR**: my fault, createPortal bubble events to parent react element

**Comment by bluebill1049** [comment]
> @bluebill1049 @xiaoyu-tamu ok I think **xiaoyu's solution is right** because I am using **createPortal** to render element in modal, but for events as stated on react docs, element kinda stays where it was created, so even it doesn't looks like I have two forms in each other but I have :D
> 
> Code interpretation
> 
> ```tsx
> <Form id="main">
>     ....
>     <Form id="create-portal-to-modal">
>      ....
>     </Form>
> </Form>
> ```
> 
> results to html
> 
> ```
> <form id="main">...</form>
> <div class="modal"><form id="create-portal-to-modal">...</form></div>
> ```
> 
> But even it is in HTML out of main form, in react tree it is still in the main and submit event bubble right into main form.
> 
> ![image](https://user-images.githubusercontent.com/3399021/86911126-10e5e100-c11b-11ea-889d-155f2e6d9765.png)
> https://reactjs.org/docs/portals.html#event-bubbling-through-portals
> 
> **TLDR**: my fault, createPortal bubble events to parent react element

Thanks @tonoslav for the clarification, i have to admit that this is new to me as well. good to know. thanks @xiaoyu-tamu!

**Comment by wowczarczyk** [comment]
I stumbled upon this one myself today. My setup of also heavily relying on `useFormContext` (pretty much all my custom form fields are using it to access useForm methods) and in my parent form I have a button that opens a Material-UI's Dialog that contains another form which was submitted the form in the background. My simple fix was too add:

```<form
          {...(dataTestid ? { 'data-testid': dataTestid } : {})}
          onSubmit={(event) => {
            event.stopPropagation();
            event.preventDefault();
            handleSubmit<T>(onSubmit as any)(event);
          }}
          className={classes.form}
          noValidate
        >```

both `stopPropagation` and `preventDefault` to the onSubmit in my `Form` component that wraps the form and utilizes the `useForm` hook. It seems to work fine.

**Comment by squiaios** [comment]
Hello @bluebill1049 ,

Is it possible to add this "problem / solution" on your FAQ please ?
